experimental/dds/tree2/src/feature-libraries/editable-tree/editableTreeBinder.ts

```diff
@@ -0,0 +1,930 @@
+					if (this.matchesPath(visitPath, current)) {
+						callbacks.forEach((callback) => collected.add(callback));
+						filtered = true;
+						break;
+					}
+				}
+				return filtered;
+			});
+			if (batchEvents.length > 0) {
+				for (const callback of collected) {
+					callback({
+						type: BindingType.Batch,
+						events: batchEvents,
+					});
+				}
+			}
+			for (const event of batchEvents) {
+				const index = sortedQueue.indexOf(event);
+				assert(index >= 0, "event confirmed in the queue");
+				sortedQueue.splice(index, 1);
+			}
+		}
+		for (const callableContext of sortedQueue) {
+			switch (callableContext.type) {
+				case BindingType.Delete:
+					for (const callback of callableContext.callbacks) {
+						// eslint-disable-next-line @typescript-eslint/no-unused-vars
+						const { callbacks, ...context } = callableContext;
+						const typedContext = { ...context, type: BindingType.Delete };
+						callback(typedContext);
+					}
+					break;
+				case BindingType.Insert:
+					for (const callback of callableContext.callbacks) {
+						// eslint-disable-next-line @typescript-eslint/no-unused-vars
+						const { callbacks, ...context } = callableContext;
+						const typedContext = { ...context, type: BindingType.Insert };
+						callback(typedContext);
+					}
+					break;
+				case BindingType.SetValue:
+					for (const callback of callableContext.callbacks) {
+						// eslint-disable-next-line @typescript-eslint/no-unused-vars
+						const { callbacks, ...context } = callableContext;
+						const typedContext = { ...context, type: BindingType.SetValue };
+						callback(typedContext);
+					}
+					break;
+				default:
+					unreachableCase(callableContext);
+			}
+		}
+		this.eventQueue.length = 0;
+		return this;
+	}
+
+	public override dispose(): void {
+		this.flush();
+		super.dispose();
+	}
+}
+
+class AbstractDataBinder<
+	B extends BinderEvents,
+	V extends AbstractPathVisitor,
+	O extends BinderOptions,
+> implements DataBinder<B>
+{
+	protected readonly visitors = new Map<EditableTree, V>();
+	protected readonly visitorLocations = new Map<V, UpPath>();
+	protected readonly unregisterHandles: Set<() => void> = new Set();
+	public constructor(
+		protected readonly options: O,
+		protected readonly visitorFactory: (anchor: EditableTree) => V,
+	) {}
+
+	public register<K extends keyof Events<B>>(
+		anchor: EditableTree,
+		eventType: K,
+		eventTrees: BindTree[],
+		listener: B[K],
+	): void {
+		// TODO: validate BindPath semantics against the schema
+		let visitor = this.visitors.get(anchor);
+		if (visitor === undefined) {
+			visitor = this.visitorFactory(anchor);
+			this.visitors.set(anchor, visitor);
+			this.unregisterHandles.add(
+				anchor[on]("subtreeChanging", (upPath: UpPath) => {
+					assert(visitor !== undefined, "visitor expected to be defined");
+					if (!this.visitorLocations.has(visitor)) {
+						this.visitorLocations.set(visitor, upPath);
+					}
+					return visitor;
```

Comment:
Will this invoke the visitor?

Reply:
This will [register](https://github.com/microsoft/FluidFramework/blob/50120aa466280c82955d43e77b793dad716ff56e/experimental/dds/tree2/src/core/tree/anchorSet.ts#L122) a visitor to be invoked during anchor set [delta traversal](https://github.com/microsoft/FluidFramework/blob/50120aa466280c82955d43e77b793dad716ff56e/experimental/dds/tree2/src/core/tree/anchorSet.ts#L618).

Reply:
ok